intermediate-goals.Rmd

```diff
@@ -0,0 +1,148 @@
+# Intermediate Outline {#intermediate-outline}
+
+This outline describes the questions that the intermediate course will answer.
+It assumes learners have completed the novice course, 
+which means they know:
+
+-   how to write code (in Python or R) that reads, analyses and plots a dataset,
+-   how to write and run unit tests,
+-   how to publish a static website using GitHub Pages,
+-   how to organize the files in a small project,
+-   the basics of licensing, and
+-   the basics of good behavior.
+
+## The Unix Shell
+
+Based on <http://swcarpentry.github.io/shell-novice/>
+and <https://merely-useful.github.io/still-magic/en/shell.html>
+
+-   How do I navigate my filesystem?
+    -   moving about (cd, ls, ., ..)
+    -   files and directories (ls, cp, mkdir, mv, rm)
+    -   paths (absolute, relative, special) 
+-   How do I examine and/or edit a text file?
+    -   difference between text and binary files
+    -   running a text editor
+    -   operating on text (cat, wc, head, tail, etc)
+    -   finding things (find, grep)
+-   How do I combine multiple commands?
+    -   pipes and filters
+    -   loops
+-   Variables
+    -   define a variable
+    -   use a variable
+    -   what's with all the quotes?
+-   Shell scripts
+    -   read/write/execute file permissions
+    -   `#!` lines
+    -   using command-line arguments in scripts
+-   Shortcuts
+    -   tab completion
+    -   previous commands (up arrow, control-r)
+    -   history
+-   Getting help
+    -   --help, man
+-   Working remotely
+    -   scp and ssh
+    -   key pairs
+
+## Automating analyses
+
+Based on <http://swcarpentry.github.io/make-novice/>
+and <https://merely-useful.github.io/still-magic/en/automate.html>
+
+-   How do I write a simple Makefile? 
+-   How can I abbreviate the rules in my Makefiles? 
+-   How can I write a Makefile to update things when my scripts have changed rather than my input files?
+-   How can I define rules to operate on similar files? 
+    -   Pattern rules
+-   How can I eliminate redundancy in my Makefiles?
+    -   Variables
+    -   Functions
+-   How should I document a Makefile?
+-   What other tools can I use?
+    -   drake for R
+    -   snakemake for Python
+
+## Reusable software
+
+Based on <https://merely-useful.github.io/still-magic/en/configure.html>
+and <https://merely-useful.github.io/still-magic/en/logging.html>
+
+-   How do I parse the command line from within a Python script?
+    -   basic structure of python scripts
+    -   argparse
+-   How would I implement a configuration file?
+    -   as a data structure (JSON)
+    -   as key/value pairs (YAML)
+-   How would I report my program's progress?
+    -   logging levels
+    -   logging channels
+    -   configuration
+-   How do I document my programs?
+    -   documentation generators
+    -   docstrings in Python
+    -   roxygen in R
```

Comment:
I would suggest even including something about how to make more fleshed out "vignettes" (in R) and conversion to websites (pkgdown in R), which would integrate the lessons in creating websites, version control (via GitHub Pages), and potentially continuous integration.

Reply:
Good suggestion. I've added reference to Sphinx / Read the Docs for Python and pkgdown for R.